### ISSUE TITLE
API to list APIs

### ISSUE THREAD

**Issue by kyyberi** [Digipalvelutehdas/APIKA#15]
As developer I want to be able to list APIs from catalogue via REST API. Meta information needed: 
- Name
- Organisation
- URL for profile in APIKA catalogue
- description
- Logo url 
- tags/keywords

**Comment by kyyberi** [comment]
Could consider making GraphQL based API?

**Comment by kyyberi** [comment]
In addition the API should have diff method to get changed metadata for APIS.

**Comment by kyyberi** [comment]
API spec must be in Swagger 2.0 format and put into Swaggerhub

**Comment by kyyberi** [comment]
Used data models could be added if possible to https://iow.csc.fi

**Comment by Nazarah** [comment]
@kyyberi hei! 
* would you like to see a "filter by organization" option in API list page?
* What do you think about idea of having an organization list view as well, where you can see APIs hosted by a particular organization and link to their profile
* from API list page, would you like to have link to the organization profile that owns a particular API?

**Comment by philippeluickx** [comment]
We could also publish the API of APIs in APIKA itself?
The diff requirement should maybe go in a different ticket though?

**Comment by jykae** [comment]
@kyyberi What you mean with this: "In addition the API should have diff method to get changed metadata for APIS." There's tool to compare REST API response diffs https://www.npmjs.com/package/node-rest-diff, though can't see why we need to track metadata changes. For that we would need versioning first.

**Comment by jykae** [comment]
@bajiat My proposal is to use Restivus (https://github.com/kahmali/meteor-restivus) for implementation with excellent https://github.com/apinf/restivus-swagger plugin that generates valid Swagger 2.0

If I am able to interpret requirements here, we could start with GET endpoints first.
Authentication is important then for POST, PUT, DELETE where we modify data.
Restivus provides authentication, https://github.com/kahmali/meteor-restivus#default-authentication can also be overridden if different needs, https://github.com/kahmali/meteor-restivus#auth

Initial GET endpoints:
GET /apis 
- List of APIs, could use query param ?org="Digipalvelutehdas"
GET /apis/:id
- Get single API by id

Currently we dont support tagging, as mentioned in description https://github.com/Digipalvelutehdas/APIKA/issues/15#issue-179076813 but other mentioned data should be in data models.

**Comment by jykae** [comment]
Support for tags seems to have separate issue: https://github.com/Digipalvelutehdas/APIKA/issues/25

**Comment by jykae** [comment]
Starting with plan above.

**Comment by kyyberi** [comment]
When do I see Swagger based design driven API development? :)

**Comment by jykae** [comment]
@kyyberi I found that using Restivus with Swagger plugin allows to develop & design in parallel which I feel effective. However if it's requirement to do this in design-driven approach, I appreciate it, use that method and try to make it in more general Node.js manner.

**Comment by jykae** [comment]
@kyyberi What format https://iow.csc.fi takes in? Data models are defined in using Meteor simple-schema in our application.

**Comment by jykae** [comment]
@kyyberi @bajiat I stick to proposed & known method for now as it's simplifies also validation as we can use existing collection permissions & schemas easily. It's no problem to add Swagger document to Swaggerhub after I have GET endpoints & we can discuss and improve API agile way. If there's proposal & tools for better approach for given time limits I will consider the option. 🙇
